enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,375 @@
+    // inject a third-party managed secret containing the certificate that 
+    // is being served on the route.
+    RouteCertificateManagedType RouteIngressConditionType = "Managed"
+
+)
+
+```
+
+#### Variation
+
+##### Alternative to the tls-secret-reference annotation
+
+***Note***: This idea has been dropped in favor of a secret reference through an annotation.
+
+The alternative to the annotation was to provide a new API field in the Route API,
+through which the user could provide the certificate reference.
+
+```go
+type CertificateReference struct {
+	// certificate resource name.
+	// +optional
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+}
+
+Type RouteSpec struct {
+	// ...
+
+	CertificateRef CertificateReference `json:"certificateRef,omitempty" protobuf:"bytes,5,opt,name=certificateRef"`
+}
+```
+
+The reasoning for the field was mainly for future proofing the API to handle integrations
+with other third-party certificate management systems. But this API introduces inconsistency
+and does not fare well when the user wants to Ingress over directly creating routes.
+
+#### Alternative to new status condition types
+
+An alternative to introducing 3 new condition types, would be to use `metadata.managedFields` 
+on Route to denote to the user which fields are managed by the secret-injector controller and 
+server-side apply would be used by the controller during updates to ensure that `.spec.tls.certificate` 
+and `.spec.tls.secret` are always managed by the controller as long as the annotation is present.
+
+This although helps in avoiding addition of new condition types will result in a few inconsistencies,
+
+- When transitioning from using managed certificates to manually managing custom certificates,
+  the `.managedFields` will still depict `.spec.tls` as managed by the controller since the update 
+  operation done by the user is only on the `.metadata.annotations`. 
+- Also since `oc` defaults to `--server-side=false`, mixing CSA and SSA results in ambiguous `.managedFields`.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+The new controller (secret-injector) will be responsible for watching routes and processing those
+that have the valid annotation for `tls-secret-reference`. This controller will also update
+`.status` of the route to contain the latest information on the certificate in use.
+
+In scenarios where the user deletes the annotation, it is the controller's 
+responsibility to reset `.spec.tls` and this will be driven based on `.status` i.e. 
+if the status denotes `RouteCertificateManagedType` and since the required annotation is 
+not present this results in the `.spec.tls` being cleared out and the `.status` updated 
+to `RouteCertificateDefaultType`.
+
+If the user deletes the secret associated with the Route through the `route.openshift.io/tls-secret-reference`
+annotation, the secret-injector controller will delete this annotation automatically and publish 
```

Comment:
Same question as in the workflow section: how exactly the new controller will know about the secret removal? A dedicated watch? How to link the secret back to the route?

Reply:
We will have a secret informer, that will trigger a re-sync of all routes in the same namespace as that of the secret. This logic is very similar to what is already present in route-controller-manager ([ref](https://github.com/openshift/route-controller-manager/blob/4d25295dbf9821107fa9adeb968220877b3331e6/pkg/route/ingress/ingress.go#L205)).

Reply:
it's sub-optimal but would do indeed.

Reply:
Updated with the details.